normandy/base/middleware.py

```diff
@@ -117,3 +122,150 @@ def process_request(self, request):
+    5. Turn the email into a Django User model instance.
+    6. If possible, extracts the first- and last name and updates the User instance persistently.
+    7. Sets that user as the user of the request thus enabling `request.user.is_active`.
+
+    NOTE! When the bearer token is present, it does a network request to the OIDC provider
+    every single time. There is no caching. Even for identical access tokens used
+    frequently. The justification for this at the Normandy Admin UI requests are
+    not high-traffic to performance critical. Also, we don't have a cache framework
+    set up.
+    """
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+        self.oidc_user_endpoint = settings.OIDC_USER_ENDPOINT
+        self.UserModel = get_user_model()
+
+    def __call__(self, request):
+        response = self.process_request(request)
+        if not response:
+            response = self.get_response(request)
+        return response
+
+    def process_request(self, request):
+        header_value = request.META.get('HTTP_AUTHORIZATION')
+        if not header_value:
+            # That's cool. If it's not there, it's not there.
+            # This middleware will not be able to sign in the user
+            # for the duration of this request.
+            return
+        matched = re.match(
+            # Note thet re.match requires it to match from the beginning
+            # otherwise it might match `Something Bearer deadbeef`
+            r'Bearer\s+(.*)',
+            header_value,
+            re.I
+        )
+        if not matched:
+            # The header was there but it didn't have a conformant "Bearer token value".
+            raise PermissionDenied('Invalid Authorization header value')
+
+        access_token = matched.group(1)
+
+        user_profile = self._fetch_oidc_email(access_token)
+        email = user_profile.get('email')
+        if not email:
+            # This would happen if someone has requested an access token
+            # from their OIDC provider *without the 'email' scope*.
+            # For example, if you do:
+            #
+            # new auth0.WebAuth({
+            #   domain: OIDC_DOMAIN,
+            #   clientID: OIDC_CLIENT_ID,
+            #   responseType: 'token id_token',
+            #   scope: 'openid profile',       <-- note lack of 'email'
+            # });
+            #
+            # Ideally the client should review their config. The
+            # access token did work but not sufficient for our needs.
+            raise PermissionDenied("User profile lacks 'email' scope.")
+
+        # Minimal normalization. Also forces all emails to be lower case
+        # which makes the UserModel lookup by email faster.
+        email = email.strip().lower()
+
+        # Awesome! Turn this email into a Django User instance.
+        try:
+            user = self.UserModel.objects.get(email=email)
```

Comment:
To me this is the crux of the authn vs authz debate. Because this code will never deny the user *after this point*, I don't think authz is happening. To me authz can only happen once you have decided who the user is, and aren't doing anything like that after this point in this middleware. (I don't think we need any changes from this comment, it was just a good visual aid for my point).

Reply:
Funny, I heard another interesting argument about the authn vs authz bikeshed today. **Authn is global and authz is when there's an action.**

Going to the will-call booth to get your ticket is authn. But actually walking through the security gate, showing your ticket, is authz. The action is "Walking into the secure area". 

Arguably, calling `GET /api/v2/service-info` is not action but `POST -d <data> /api/v2/recipes` is.

Reply:
Another argument for calling it authz here in this PR is that it's stateless. Authn is usually state-full. I.e. turning credentials into something that lasts, e.g. a session cookie, or an access token in localStorage.

Reply:
Do you want me to search-and-replace authz for authn? I'm not wedded to either. The fact that it arose a "debate" by several experienced developers means it's tricky notion to get right.